src/ServiceControl.Monitoring/QueueLength/QueueLengthCalculator.cs

```diff
@@ -3,45 +3,80 @@
     using System;
     using System.Collections.Concurrent;
     using System.Collections.Generic;
+    using System.Linq;
 
     class QueueLengthCalculator : IQueueLengthCalculator
     {
-        public void UpdateReceivedSequence(string key, long value, string queue)
+        public bool UpdateReceivedSequence(string endpointName, string key, double value, string queue)
         {
-            receivedSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
-            sessionKeyToQueueMapping.GetOrAdd(key, _ => queue);
+            var virtualQueue = new VirtualQueueId
+            {
+                EndpointName = endpointName,
+                QueueName = queue,
+                SessionKey = key
+            };
+
+            receivedSequences.AddOrUpdate(virtualQueue, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return sentSequences.ContainsKey(key);
         }
 
-        public void UpdateSentSequence(string key, long value)
+        public List<VirtualQueueId> UpdateSentSequence(string key, double value)
         {
-            sentSequences.AddOrUpdate(key, _ => value, (_, prev) => Math.Max(prev, value));
+            sentSequences.AddOrUpdate(key, _ => value, (_, previousValue) => Math.Max(previousValue, value));
+
+            return receivedSequences.Where(kv => kv.Key.SessionKey == key).Select(kv => kv.Key).ToList();
         }
 
-        public Dictionary<string, long> GetQueueLengths()
+        public Dictionary<VirtualQueueId, double> GetQueueLengths()
         {
-            var queueLengths = new Dictionary<string, long>();
+            var queueLengths = new Dictionary<VirtualQueueId, double>();
 
-            foreach (var kvp in sessionKeyToQueueMapping)
+            foreach (var receivedSequence in receivedSequences)
             {
-                var queueName = kvp.Value;
-                long queueLength;
-                queueLengths.TryGetValue(queueName, out queueLength);
-
-                long sent;
-                sentSequences.TryGetValue(kvp.Key, out sent);
-                var received = receivedSequences[kvp.Key];
+                double sent;
+                sentSequences.TryGetValue(receivedSequence.Key.SessionKey, out sent);
 
-                queueLength += sent;
-                queueLength -= received;
+                var received = receivedSequence.Value;
 
-                queueLength = Math.Max(0, queueLength);
-                queueLengths[queueName] = queueLength;
+                queueLengths.Add(receivedSequence.Key, Math.Max(0, sent - received));
             }
+
```

Comment:
It might be possible to replace this whole method with a single LINQ statement. Something like

```
(from r in receivedSequences
join s in sentSequeneces on r.Key.SessionKey equals s.Key
select new KeyValuePair<>(r.Key, Math.Max(s?.Value ?? 0 -r.Value, 0)).ToDictionary();
```